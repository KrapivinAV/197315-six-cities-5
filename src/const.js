export const OfferListStyle = {
  MAIN_SCREEN: `cities__places-list places__list tabs__content`,
  PROPERTIES_SCREEN: `near-places__list places__list`,
  FAVORITES_SCREEN: `favorites__places`
};

export const OfferCardStyleSet = {
  MAIN_SCREEN: {
    ARTICLE: `cities__place-card`,
    IMAGE_WRAPPER: `cities__image-wrapper`,
    INFO: ``
  },
  PROPERTIES_SCREEN: {
    ARTICLE: `near-places__card`,
    IMAGE_WRAPPER: `near-places__image-wrapper`,
    INFO: ``
  },
  FAVORITES_SCREEN: {
    ARTICLE: `favorites__card`,
    IMAGE_WRAPPER: `favorites__image-wrapper`,
    INFO: `favorites__card-info`
  }
};

export const offerTypes = {
  apartment: `Apartment`,
  room: `Private Room`,
  house: `House`,
  hotel: `Hotel`
};

export const cities = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`
];

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const AppRoute = {
  MAIN: `/`,
  LOGIN: `/login`,
  FAVORITES: `/favorites`,
  PROPERTIES: `/offer/:id`
};

export const NEAR_OFFERS_MAX_QUANTITY = 3;
export const REVIEWS_MAX_QUANTITY = 10;
