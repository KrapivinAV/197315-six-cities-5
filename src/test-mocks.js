export const offerFirstCityIsFavoriteIsPremiumHostIsPro = {
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`
  ],
  title: `The house among olive`,
  rating: 3,
  type: `apartment`,
  bedrooms: 5,
  price: 226,
  goods: [
    `Dishwasher`,
    `Fridge`,
    `Towels`
  ],
  host: {
    id: 25,
    name: `Angelina`,
    avatarUrl: `/img/avatar-angelina.jpg`,
    isPro: true
  },
  description: `Peaceful studio in the most wanted area in town. Quiet house Near of everything.`,
  location: {
    latitude: 48.87761,
    longitude: 2.333499,
    zoom: 16
  },
  id: 1,
  isFavorite: true,
  isPremium: true,
  maxAdults: 8
};

export const offerFirstCityNotFavoriteNotPremiumHostNotPro = {
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`
  ],
  title: `The house among olive`,
  rating: 3,
  type: `apartment`,
  bedrooms: 5,
  price: 226,
  goods: [
    `Dishwasher`,
    `Fridge`,
    `Towels`
  ],
  host: {
    id: 25,
    name: `Angelina`,
    avatarUrl: `/img/avatar-angelina.jpg`,
    isPro: false
  },
  description: `Peaceful studio in the most wanted area in town. Quiet house Near of everything.`,
  location: {
    latitude: 48.87761,
    longitude: 2.333499,
    zoom: 16
  },
  id: 2,
  isFavorite: false,
  isPremium: false,
  maxAdults: 8
};

export const offerSecondCityIsFavoriteNotPremiumHostNotPro = {
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`
  ],
  title: `The house among olive`,
  rating: 3,
  type: `apartment`,
  bedrooms: 5,
  price: 226,
  goods: [
    `Dishwasher`,
    `Fridge`,
    `Towels`
  ],
  host: {
    id: 25,
    name: `Angelina`,
    avatarUrl: `/img/avatar-angelina.jpg`,
    isPro: false
  },
  description: `Peaceful studio in the most wanted area in town. Quiet house Near of everything.`,
  location: {
    latitude: 48.87761,
    longitude: 2.333499,
    zoom: 16
  },
  id: 3,
  isFavorite: true,
  isPremium: false,
  maxAdults: 8
};

export const REVIEW_FIRST = {
  id: 1,
  rating: 3,
  comment: `99999999999999999999999999999999999999999999999999999`,
  date: `2020-11-30T06:28:10.586Z`,
  user: {
    id: 1,
    name: `ak`,
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/1.jpg`,
    isPro: true
  }
};

export const REVIEW_SECOND = {
  id: 2,
  rating: 5,
  comment: `99999999999999999999999999999999999999999999999999999`,
  date: `2020-10-30T06:28:10.586Z`,
  user: {
    id: 1,
    name: `ak`,
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/1.jpg`,
    isPro: false
  }
};
