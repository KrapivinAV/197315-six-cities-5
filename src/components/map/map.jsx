import React, {Component} from "react";
import PropTypes from "prop-types";
import PropTypesSet from "../../prop-types-set";
import {connect} from "react-redux";
import "leaflet/dist/leaflet.css";
import leaflet from "leaflet";
import {ScreenMarker} from "../../const";

class Map extends Component {
  constructor(props) {
    super(props);

    this.map = null;
  }

  componentDidMount() {
    const {offers} = this.props;

    const icon = leaflet.icon({
      iconUrl: `/img/pin.svg`,
      iconSize: [30, 30]
    });

    this.map = leaflet.map(`map`, {
      center: [offers[0].city.location.latitude, offers[0].city.location.longitude],
      zoom: offers[0].city.location.zoom,
      zoomControl: false,
      marker: true
    });

    this.map.setView([offers[0].city.location.latitude, offers[0].city.location.longitude], offers[0].city.location.zoom);

    leaflet
    .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    })
    .addTo(this.map);

    offers.forEach((offer) => {
      const offerCoordinates = [
        offer.location.latitude,
        offer.location.longitude
      ];

      leaflet
        .marker(offerCoordinates, {icon})
        .addTo(this.map);
    });
  }

  componentDidUpdate() {
    const {offers, currentOfferCardId, screenMarker} = this.props;

    this.map.eachLayer((layer) => {
      if (layer.options.icon) {
        layer.remove();
      }
    });

    offers.forEach((offer) => {
      const offerCoordinates = [
        offer.location.latitude,
        offer.location.longitude,
      ];

      const icon = leaflet.icon({
        iconUrl: (screenMarker === ScreenMarker.MAIN && +currentOfferCardId === offer.id) || (screenMarker === ScreenMarker.PROPERTIES && offers[0].id === offer.id) ? `/img/pin-active.svg` : `/img/pin.svg`,
        iconSize: [30, 30]
      });

      leaflet
        .marker(offerCoordinates, {icon})
        .addTo(this.map);
    });

    this.map.setView([offers[0].city.location.latitude, offers[0].city.location.longitude], offers[0].city.location.zoom);
  }

  render() {
    return (
      <div
        id="map"
        style={{height: `100%`}}>
      </div>
    );
  }
}

Map.propTypes = {
  offers: PropTypes.arrayOf(PropTypesSet.offer).isRequired,
  currentOfferCardId: PropTypes.string.isRequired,
  screenMarker: PropTypes.string.isRequired
};

const mapStateToProps = ({STATE}) => ({
  currentOfferCardId: STATE.currentOfferCardId
});

export {Map};
export default connect(mapStateToProps)(Map);
