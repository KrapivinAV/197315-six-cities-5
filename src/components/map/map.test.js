import React from "react";
import renderer from "react-test-renderer";
import {Map} from "./map";
import {Provider} from "react-redux";
import {Route, BrowserRouter} from "react-router-dom";
import configureStore from "redux-mock-store";
import {ScreenMarker} from "../../const";
import {OFFER_FIRST_CITY_IS_FAVORITE_IS_PREMIUM_HOST_IS_PRO, OFFER_FIRST_CITY_NOT_FAVORITE_NOT_PREMIUM_HOST_NOT_PRO} from "../../test-mocks";

const CURRENT_OFFER_CARD_ID = `1`;

jest.mock(`leaflet`, () => ({
  icon: () => {},
  eachLayer: () => {},
  map: () => ({
    setView: () => {}
  }),
  tileLayer: () => ({
    addTo: () => {}
  }),
  marker: () => ({
    addTo: () => {}
  })
}));

describe(`Render Map`, () => {
  it(`On Main Screen`, () => {

    const mockStore = configureStore([]);
    const store = mockStore({});

    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <Route>
                <Map
                  offers={[OFFER_FIRST_CITY_IS_FAVORITE_IS_PREMIUM_HOST_IS_PRO, OFFER_FIRST_CITY_NOT_FAVORITE_NOT_PREMIUM_HOST_NOT_PRO]}
                  currentOfferCardId={CURRENT_OFFER_CARD_ID}
                  screenMarker={ScreenMarker.MAIN}
                />
              </Route>
            </BrowserRouter>
          </Provider>,
          {
            createNodeMock: () => {
              const div = document.createElement(`div`);
              div.id = `map`;
              return div;
            }
          }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`On Properties Screen`, () => {

    const mockStore = configureStore([]);
    const store = mockStore({});

    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <Route>
                <Map
                  offers={[OFFER_FIRST_CITY_IS_FAVORITE_IS_PREMIUM_HOST_IS_PRO, OFFER_FIRST_CITY_NOT_FAVORITE_NOT_PREMIUM_HOST_NOT_PRO]}
                  currentOfferCardId={CURRENT_OFFER_CARD_ID}
                  screenMarker={ScreenMarker.PROPERTIES}
                />
              </Route>
            </BrowserRouter>
          </Provider>,
          {
            createNodeMock: () => {
              const div = document.createElement(`div`);
              div.id = `map`;
              return div;
            }
          }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

});
