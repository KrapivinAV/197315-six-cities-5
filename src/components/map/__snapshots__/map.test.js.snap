// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Map On Main Screen 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;

exports[`Render Map On Properties Screen 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;
